### PR TITLE
[build-script] Add option to build lld as part of LLVM

### PR DESCRIPTION
**Pull request by ahoppen**
I find that lld links faster than the default linker for release builds and it’s convenient if we can just build it as part of Swift’s LLVM build.

Note that to use lld, you still need to set `extra-cmake-options=-DLLVM_ENABLE_LLD:BOOL=TRUE` manually. This PR does not do so because we can only start using that flag *after* lld has been built.

### CODE REVIEW

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Comment by ahoppen** [comment]
@swift-ci Please smoke test macOS

**Comment by gottesmm** [comment]
@ahoppen I thought we already built LLD by default on certain platforms. Specifically Linux/windows. Or is me memory wrong. @compnerd

File: utils/swift_build_support/swift_build_support/build_script_invocation.py [hunk 1]
```diff
@@ -429,6 +429,11 @@ def convert_to_impl_arguments(self):
                 "--llvm-install-components=%s" % args.llvm_install_components
             ]
 
+        if args.build_lld:
+            impl_args += [
+                "--llvm-install-components=%s" % args.llvm_install_components
```
Comment (gottesmm) [inline utils/swift_build_support/swift_build_support/build_script_invocation.py:434]: Does llvm install components accumulate in build-script like tthis? Doesn't this break the usage of llvm-install-components above.

File: utils/swift_build_support/swift_build_support/build_script_invocation.py [hunk 2]
```diff
@@ -429,6 +429,11 @@ def convert_to_impl_arguments(self):
                 "--llvm-install-components=%s" % args.llvm_install_components
```
Comment (gottesmm) [inline utils/swift_build_support/swift_build_support/build_script_invocation.py:429]: This specifically.

File: utils/build-script-impl [hunk 3]
```diff
@@ -1832,7 +1833,7 @@ for host in "${ALL_HOSTS[@]}"; do
                 #
                 # This makes it easier to build target stdlibs on systems that
                 # have old toolchains without more modern linker features.
-                if [[ "$(uname -s)" != "Darwin" ]] ; then
+                if [[ "$(uname -s)" != "Darwin" || "${BUILD_LLD}" ]] ; then
                     if [[ ! "${SKIP_BUILD_LLD}" ]]; then
```
Comment (gottesmm) [inline utils/build-script-impl:1837]: @ahoppen I was thinking... we already have a skip-build-lld option... so we really shouldn't add another build-lld option to build-script-impl. In general we want to simplify build-script-impl by not adding new state like this.

I have a counter-proposal, why don't we remove the Darwin check here. Instead, in build-script lets just pass in skip-build-lldb to build-script-impl if we are on Darwin and build-lld isn't set. Then build-script-impl is simpler and doesn't have an additional variable. What are your thoughts?

File: utils/swift_build_support/swift_build_support/build_script_invocation.py (hunk 1 above)
Comment (ahoppen) [reply utils/swift_build_support/swift_build_support/build_script_invocation.py:434]: Not sure how I managed to commit this, but this should have been 
```
if args.build_lld:
  impl_args += ["--build-lld"]
```
🤦

**Comment by ahoppen** [comment]
Very good idea, @gottesmm. I changed the PR to remove the Darwin check in build-script-impl and instead make build-script decide whether lld should be built.

**Comment by ahoppen** [comment]
@swift-ci Please smoke tet

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: utils/build-script-impl [hunk 4]
```diff
@@ -1826,16 +1826,8 @@ for host in "${ALL_HOSTS[@]}"; do
...
-                # compile the stdlib.
-                #
-                # This makes it easier to build target stdlibs on systems that
-                # have old toolchains without more modern linker features.
-                if [[ "$(uname -s)" != "Darwin" ]] ; then
-                    if [[ ! "${SKIP_BUILD_LLD}" ]]; then
-                        llvm_enable_projects+=("lld")
-                    fi
+                if [[ ! "${SKIP_BUILD_LLD}" ]]; then
+                    llvm_enable_projects+=("lld")
```
[the remark this answers is not in the document]
Comment (ahoppen) [reply utils/build-script-impl:1830]: What do you mean by a build-script unit test? I don’t think I ever saw unit tests for `build-script` 🤔

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

File: utils/build-script-impl (hunk 4 above)
[the remark this answers is not in the document]
Comment (ahoppen) [reply utils/build-script-impl:1837]: I added a unit test. @gottesmm Could you take another look?

**Comment by ahoppen** [comment]
@swift-ci Please smoke test Linux

**Comment by ahoppen** [comment]
@swift-ci Please smoke test

**Review by gottesmm — APPROVED** [review]
LGTM. Thanks for adding the test!

**Pull request merged**